### ISSUE TITLE
Thrown errors in Promise within $beforeInsert causes unhandled rejection error

### ISSUE THREAD

**Issue by jzumbrun** [Vincit/objection.js#30]
I have this query now in a controller:

```
User.query()
		.where('email', req.body.email)
		.first()
		.then(function(email){
			if(email){
				throw {data: {email: 'Already exists.'}};
			}
			return User.query().insert(req.body);
		})
		.then(function (user) {
			res.send({ token: user.token() });
		})
		.catch(function(error){
			console.log('error', error);
			return res.status(422).send(error.data || error);
		});
```

I would like to have this instead:

```
User.query()
			.insert(req.body)
			.then(function (user) {
				res.send({ token: user.token() });
			}).catch(function(error){
				return res.status(422).send(error);
			});
```

and then in `$beforeInsert` have the email checking query:

```
self.$beforeInsert = function(){

			if(!validatePassword(self.password)){
				// THIS THROW WORKS!
				throw {data: {password: 'String is too short (' + self.password.length + ' chars), minimum 8'}};
			}

			User.query()
			.where('email', self.email)
			.first()
			.then(function(user){
				console.log(user);
				if(user){
					// THIS THROW CAUSES BLUEBIRD UNHANDLED REJECTION ERROR 
					throw {data: {email: 'Already exists.'}};
				}

				// Create salt and hash password
				self.salt = new Buffer(crypto.randomBytes(16).toString('base64'), 'base64');
				self.password = hashPassword(self.password);
			})
			.catch(function(error){
				console.log(error);
				throw {data: {error: error}};
			});
		};
```

The first throw in `$beforeInsert` works. However the second throw, the one in a now "nested" promise causes an unhandled rejection error, most likely due to the fact that the promise in `self.$beforeInsert` does not return iteself back to the `then` method that handles the first insert query.

Is there a way to get a promise made in `$beforeInsert` to return back to the initial insert promise and have the errors caught at the parent level so bluebird is not complaining about unhandled rejection errors?

Thanks - Jon

**Comment by koskimas** [comment]
Currently the `$beforeInsert`, `$beforeUpdate` and friends are called synchronously and you cannot do async stuff in them. I actually thought of this when I added the hooks, but decided not to make them async.

I will make them async in the next update as it is an easy change and doesn't break anything. You will have to return the `User.query().where('email', self.email)` promise from the handler after the fix to make this work.

One thing caught my eye in the code though: Why are you doing `self.$beforeInsert = function ()...` instead of doing `User.prototype.$beforeInsert = function ()...` where you create the `Model` class?

**Comment by koskimas** [comment]
I spoke too soon. I tried to make the `$before` methods asynchronous, but realised that I would have to make pretty big changes to the architecture and lots of breaking changes. I think it is better to keep the `$before` methods synchronous after all.

I'm working on a solution to your propblem though. After the next update there are couple of ways to do what you want. The first one is simple, but you have to use a custom `insert` method:

```js
// Add the custom `insert` method to your User model. You have to use this instead of
// User.query().insert() when you insert users.
User.insert = function (user) {
  var self = this; 

  // Always use `this` instead of the model class directly. This ensures that the queries
  // take part in the transaction, if `this` is bound to one.
  var insertQuery = this.query().insert(user);

  // Register a function to be called before the query is built and executed. You
  // can return promises from these and chain them just like with a `then` method
  // of a promise.
  return insertQuery.runBefore(function () {
    return self.query().where('email', user.email).first();
  }).runBefore(function (userByEmail) {
    if (userByEmail) {
      throw {data: {email: 'Already exists.'}};
    }
    // Create salt and hash password
    user.salt = new Buffer(crypto.randomBytes(16).toString('base64'), 'base64');
    user.password = hashPassword(self.password);
  });
};
```

The second option requires more code, but you don't have to use a custom `insert` method:

```js
var QueryBuilder = require('objection').QueryBuilder;
var Model = require('objection').Model;

// Create a subclass of QueryBuilder.
function UserQueryBuilder() {
  QueryBuilder.apply(this, arguments);
}

QueryBuilder.extend(UserQueryBuilder);

// Override the `insert` method.
UserQueryBuilder.prototype.insert = function (user) {
  // You should access the model class through `modelClass()` method of the
  // QueryBuilder. You could use `User` directly, but that wouldn't work correctly
  // with transactions.
  var User = this.modelClass();

  // Register a function to be called before the query is built and executed. You
  // can return promises from these and chain them just like with a `then` method
  // of a promise.
  this.runBefore(function () {
    return User.query().where('email', user.email).first();
  }).runBefore(function (userByEmail) {
    if (userByEmail) {
      throw {data: {email: 'Already exists.'}};
    }
    // Create salt and hash password
    user.salt = new Buffer(crypto.randomBytes(16).toString('base64'), 'base64');
    user.password = hashPassword(self.password);
  });
  
  // Call the original `insert` method.
  return QueryBuilder.prototype.insert.call(this, user);
};

// This is your User model.
function User() {
  Model.apply(this, arguments);
}

Model.extend(User);

// Use the UserQueryBuilder instead of QueryBuilder.
User.QueryBuilder = UserQueryBuilder;

// ... More methods and stuff.

// Now you can just do this:
User.query().insert(req.body).then(...);
```

**Comment by jzumbrun** [comment]
I like the `User.insert()` option better. 

I know objection.js is an "extension" to knex so using `User.query()` makes sense but I would rather use the abstraction layer of `User.insert()` anyways as it feels more ORM/ODM-ish. 

I would like to see objection.js provide `User.insert()` without having to define it myself including the `$beforeInsert()` called within that `User.insert()` while having it pass the query object to `$beforeInsert(query)` so we can just use it like this:

```
User.prototype.$beforeInsert(query) = function {
var self = this;
return query.runBefore(function () {
    return self.query().where('email', self.email).first();
  }).runBefore(function (userByEmail) {
    if (userByEmail) {
      throw {data: {email: 'Already exists.'}};
    }
    // Create salt and hash password
    self.salt = new Buffer(crypto.randomBytes(16).toString('base64'), 'base64');
    self.password = hashPassword(self.password);
  });
};
```

And then just use:

```
User.insert()
 .then(function (user) {
                res.send({ token: user.token() });
            }).catch(function(error){
                return res.status(422).send(error);
            });
```

Which would mean all the other crud methods would be abstracted to:

```
User.insert()...
User.update()...
User.patch()...
User.delete()...

etc...
```

But that is just my opinion :smile:

**Comment by koskimas** [comment]
Passing the query builder to the `$beforeInsert` method requires the same architecture changes as making the `$beforeInsert` asynchronous. But you are right that the only clean solution here is to do the "before stuff" in the `$beforeInsert` method. I will do the larger refactoring and make it possible. a User model that does stuff like this is a very common thing to do.

So after the refactor you will be able to do this:

```js
User.prototype.$beforeInsert = function () {
  var self = this;
  var User = this.constructor;

  return User.query().where('email', self.email).first().then(function (userByEmail) {
    if (userByEmail) {
      throw {data: {email: 'Already exists.'}};
    }
    // Create salt and hash password
    self.salt = new Buffer(crypto.randomBytes(16).toString('base64'), 'base64');
    self.password = hashPassword(self.password);
  });
};
```

The `User.insert()`, `User.update()` etc. methods are a separate issue. Feel free to create another ticket. Objection.js originally had those methods (before it was released to the public) but some people thought they were confusing. It wasn't clear where the model stopped and the query builder started. So we changed it to the current model. If the ticket gets a lot of attention I will gladly add the methods back though.
